executor/cook/launcher.py

```diff
@@ -0,0 +1,181 @@
+            stdout = open('stdout.' + str(self.name), 'w+')
+            stderr = open('stderr.' + str(self.name), 'w+')
+
+        return RunningCommand(
+            *self,
+            stdout = stdout,
+            stderr = stderr,
+            process = subprocess.Popen(
+                self.value,
+                shell = True,
+                stdout = stdout,
+                stderr = stderr,
+                env = dict(os.environ, **env)
+            )
+        )
+
+# set default values for (name, async, guard)
+Command.__new__.__defaults__ = (None, False, False, False)
+
+class RunningCommand(namedtuple(
+        'RunningCommand', Command._fields + ('stdout', 'stderr', 'process'))
+):
+    """
+    An internal representation of a running command. Using a namedtuple helps catch
+    issues with data integrity.
+    """
+
+    def kill(self):
+        self.stdout.close()
+        self.stderr.close()
+
+        if self.process.returncode is None:
+            self.process.terminate()
+            for _ in range(0, 10):
+                if self.process.poll() is not None:
+                    break
+                time.sleep(1)
+            if self.process.returncode is None:
+                self.process.kill()
+
+        return self.process.returncode
+
+    def poll(self):
+        return self.process.poll()
+
+    def is_guard(self):
+        return self.guard or False
+
+    def is_async(self):
+        return self.async or False
+
+    def is_sync(self):
+        return not self.is_async()
+
+    def is_running(self):
+        return self.process.poll() is None
+
+    def is_crashed(self):
+        return not self.is_running() and self.process.poll() is not 0
+
+def run_commands(commands, stop = None, get_env = lambda: {}, update_task = lambda *_: {}):
+    """
+    Execute a list of commands, waiting for all synchronous commands to complete. Will also
+    exit if the stop event is set.
+
+    Returns a list of exit codes for the command processes.
+    """
+    did_abort = True
+    running_commands = []
+    iterator_commands = (Command(**c) for c in commands)
+
+    try:
+        while not (stop and stop.isSet()):
+            should_await = [c for c in running_commands if c.is_running() and c.is_sync()]
+            should_abort = [c for c in running_commands if c.is_crashed() and c.is_guard()]
+
+            if should_abort:
+                break
+            elif not should_await:
+                if len(commands) == len(running_commands):
+                    did_abort = False
+                    break
+                else:
+                    running_commands.append(next(iterator_commands).run(get_env()))
+            else:
+                time.sleep(1)
+
+            update_task([c.poll() for c in running_commands], False, should_abort)
+    except Exception as e:
+        logging.exception("Exception in run_commands")
+    finally:
+        update_task([c.kill() for c in running_commands], True, did_abort)
```

Comment:
Should async commands that are still running be first sent SIGTERM before we send SIGKILL.  The old documentation indicates that was the desired behavior.